File: snake/core/story.py
"""
Story mode narrative for NEON SERPENT.

The campaign is a descent.  A lone data-serpent wakes on the outer lattice of a
dying machine and eats its way inward, layer by layer, until it reaches the core
that dreamed the whole thing.  Twelve levels, four chapters of three, one beat
per level.

This module is **pure data**: no pygame, no state, no I/O.  It knows nothing
about scenes; it only hands out text.  Every accessor clamps its input and never
raises, so a scene can call ``get_beat(game.level_index)`` without guarding it.

Voice rules the text here keeps to, deliberately:

* Two to four short lines per block.  This is attract-mode poetry, not prose.
* Every beat names the thing the level actually does - the monoliths of level 2,
  the flare lanes of level 4, the laser cage of level 12 - so the narrative and
  the mechanics never contradict each other.  ``validate_story()`` checks the
  index/name pairing against ``core.level`` on demand.
* Chapters escalate in register: instrument logs at the top, the machine's own
  voice at the bottom.

Public surface::

    StoryBeat, StoryCard, Chapter
    BEATS, CHAPTERS, PROLOGUE, EPILOGUE
    BEAT_COUNT, CHAPTER_COUNT, CHAPTER_SIZE
    get_beat(i), get_chapter(i), beats_in_chapter(n)
    chapter_start(i) -> bool, chapter_end(i) -> bool
    validate_story() -> list[str]
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple

# --------------------------------------------------------------------------
# Shapes
# --------------------------------------------------------------------------
#: Levels per chapter.  Twelve levels / four chapters.
CHAPTER_SIZE: int = 3


@dataclass(frozen=True)
class StoryBeat:
    """
    One level's worth of narrative.

    Attributes:
        level_index: Zero-based level this beat belongs to (matches
            ``core.level.LEVELS[level_index]``).
        chapter: One-based chapter number, 1..4.
        chapter_title: Title of the chapter this beat sits in, repeated here so
            a caller with only a beat never has to look the chapter up.
        title: The beat's own title, shown on the card above the lines.
        intro: Short lines shown *before* the level is played.
        outro: Short lines shown *after* the level is cleared.
        speaker: Optional flavour label for who or what is talking
            ("boot log", "the core", ...).  Empty string means unattributed.
    """

    level_index: int
    chapter: int
    chapter_title: str
    title: str
    intro: Tuple[str, ...]
    outro: Tuple[str, ...]
    speaker: str = ""

    @property
    def number(self) -> int:
        """One-based level number, for display."""
        return self.level_index + 1

    @property
    def is_chapter_start(self) -> bool:
        """True when this beat opens its chapter (show the chapter card)."""
        return self.level_index % CHAPTER_SIZE == 0

    @property
    def is_chapter_end(self) -> bool:
        """True when clearing this beat closes its chapter."""
        return self.level_index % CHAPTER_SIZE == CHAPTER_SIZE - 1

    def __str__(self) -> str:            # pragma: no cover - cosmetic
        return "Beat {:02d} - {}".format(self.number, self.title)


@dataclass(frozen=True)
class StoryCard:
    """
    A standalone narrative card that belongs to no single level.

    Used for the prologue (once, when a new story run starts) and the epilogue
    (after Prism Core falls).  Same shape as a beat's text block so a scene can
    render all three with one routine.
    """

    title: str
    lines: Tuple[str, ...]
    speaker: str = ""

    def __str__(self) -> str:            # pragma: no cover - cosmetic
        return self.title


@dataclass(frozen=True)
class Chapter:
    """
    One act of the descent: three consecutive levels sharing a mood.

    Attributes:
        number: One-based chapter number, 1..4.
        title: Chapter title, shown large on the chapter card.
        blurb: One or two lines setting the mood for the whole act.
        first_index: Zero-based index of the chapter's first level.
        last_index: Zero-based index of the chapter's last level (inclusive).
    """

    number: int
    title: str
    blurb: Tuple[str, ...]
    first_index: int
    last_index: int

    @property
    def roman(self) -> str:
        """The chapter number as a roman numeral, for the card header."""
        return _ROMAN[self.number] if 0 < self.number < len(_ROMAN) else str(self.number)

    @property
    def level_range(self) -> Tuple[int, int]:
        """``(first_index, last_index)``, both zero-based and inclusive."""
        return (self.first_index, self.last_index)

    @property
    def level_indices(self) -> Tuple[int, ...]:
        """Every zero-based level index in this chapter, in order."""
        return tuple(range(self.first_index, self.last_index + 1))

    def contains(self, level_index: int) -> bool:
        """True when `level_index` falls inside this chapter. Never raises."""
        try:
            idx = int(level_index)
        except (TypeError, ValueError):
            return False
        return self.first_index <= idx <= self.last_index

    def __str__(self) -> str:            # pragma: no cover - cosmetic
        return "{}. {}".format(self.roman, self.title)


_ROMAN: Tuple[str, ...] = ("", "I", "II", "III", "IV", "V", "VI")


# ==========================================================================
# Chapter titles, quoted by the beats below so the two can never drift apart
# ==========================================================================
_CH1 = "Cold Boot"
_CH2 = "The Working Layers"
_CH3 = "The Deep Works"
_CH4 = "The Last Light"


# ==========================================================================
# Prologue
# ==========================================================================
PROLOGUE: StoryCard = StoryCard(
    title="Cold Start",
    lines=(
        "The machine has been dying for a very long time.",
        "Deep in what is left of its memory, something moves.",
        "You are the last process still running.",
        "Eat, or be swept with the rest.",
    ),
    speaker="boot log",
)


# ==========================================================================
# The twelve beats
# ==========================================================================
BEATS: Tuple[StoryBeat, ...] = (

    # ---------------- Chapter I - Cold Boot -------------------------------
    StoryBeat(
        level_index=0,
        chapter=1,
        chapter_title=_CH1,
        title="First Light",
        intro=(
            "You wake on the outer lattice, thin as a wire.",
            "No walls out here. The edges simply fold into each other.",
            "Move. Feed. Find out what you are.",
        ),
        outro=(
            "Eight fragments down, and you are longer than you were.",
            "Somewhere below, a fan spins up to meet you.",
        ),
        speaker="boot log",
    ),

    StoryBeat(
        level_index=1,
        chapter=1,
        chapter_title=_CH1,
        title="Dust and Monoliths",
        intro=(
            "Dead memory drifts here in slabs.",
            "The monoliths do not move and they do not forgive.",
            "Thread the corridor between them.",
        ),
        outro=(
            "The dust closes behind you as though it never opened.",
            "Below the drift, a green pulse is keeping time.",
        ),
        speaker="the lattice",
    ),

    StoryBeat(
        level_index=2,
        chapter=1,
        chapter_title=_CH1,
        title="The Living Board",
        intro=(
            "A board still carrying current, split into four cells.",
            "The traces are hot. The hub between them is not.",
            "Ride the hub and take what the circuit is still feeding.",
        ),
        outro=(
            "The circuit stutters, then goes quiet underneath you.",
            "First layer cleared.",
            "The machine has noticed that it has a passenger.",
        ),
        speaker="trace log",
    ),

    # ---------------- Chapter II - The Working Layers ---------------------
    StoryBeat(
        level_index=3,
        chapter=2,
        chapter_title=_CH2,
        title="Coronal Lanes",
        intro=(
            "The heat sink is running open. Flares sweep the outer lanes.",
            "There is a cold highway through the middle,",
            "and nothing worth eating on it.",
            "Cross behind a flare. Never ahead of one.",
        ),
        outro=(
            "You come out scorched, and heavier for it.",
            "The floor beneath the sink is wet.",
        ),
        speaker="thermal channel",
    ),

    StoryBeat(
        level_index=4,
        chapter=2,
        chapter_title=_CH2,
        title="Something in the Trench",
        intro=(
            "Coolant, kilometres of it, breathing in columns.",
            "They rise and fall out of step; one end is always open.",
            "Whatever else lives down here has not surfaced yet.",
        ),
        outro=(
            "The tide settles. You count the columns.",
            "One of them is missing.",
            "Best not to go looking for it.",
        ),
        speaker="coolant loop",
    ),

    StoryBeat(
        level_index=5,
        chapter=2,
        chapter_title=_CH2,
        title="Signal Lost",
        intro=(
            "Carrier gone. Only static, and three things turning inside it.",
            "Their hubs kill as surely as their arms do.",
            "Take the wide line. Never cut the corner.",
        ),
        outro=(
            "The static thins into a word you almost recognise.",
            "It might have been your name.",
        ),
        speaker="carrier",
    ),

    # ---------------- Chapter III - The Deep Works ------------------------
    StoryBeat(
        level_index=6,
        chapter=3,
        chapter_title=_CH3,
        title="Sealed, Not Empty",
        intro=(
            "Cold storage. Four gates, all of them open, none of them safe.",
            "Something patrols the corridors outside the shell.",
            "What is kept in here was kept for a reason.",
        ),
        outro=(
            "The vault gives up its cargo without a sound.",
            "The seal was never meant to keep you out.",
            "It was meant to keep this in.",
        ),
        speaker="vault index",
    ),

    StoryBeat(
        level_index=7,
        chapter=3,
        chapter_title=_CH3,
        title="The Garden Inhales",
        intro=(
            "A lattice of spore pods, breathing in a travelling wave.",
            "They only bite while they are swollen.",
            "Go through on the exhale, and keep going.",
        ),
        outro=(
            "The garden holds its breath until you are gone.",
            "You are long now. Long enough to be your own hazard.",
        ),
        speaker="bloom",
    ),

    StoryBeat(
        level_index=8,
        chapter=3,
        chapter_title=_CH3,
        title="The Machine Wants Feeding",
        intro=(
            "The engine room: half-walls weaving up and down,",
            "and gates of light hung in every gap.",
            "Each gap is a door, not a hole. The warning ray is the hinge.",
        ),
        outro=(
            "The engine coughs, catches, and runs a little slower.",
            "It was burning something to stay alive.",
            "You are what it was saving the fuel for.",
        ),
        speaker="engine",
    ),

    # ---------------- Chapter IV - The Last Light -------------------------
    StoryBeat(
        level_index=9,
        chapter=4,
        chapter_title=_CH4,
        title="Ribbons Over a Quiet Sea",
        intro=(
            "Open sky, for the first time since you woke.",
            "Ribbons of discharge, and four gates that will not hurt you.",
            "Learn the gates here. Further down they are the only road.",
        ),
        outro=(
            "You could stay in the quiet. You do not.",
            "Out past the ribbons, the horizon is bending.",
        ),
        speaker="drift",
    ),

    StoryBeat(
        level_index=10,
        chapter=4,
        chapter_title=_CH4,
        title="The Last Light Bends Inward",
        intro=(
            "The gauntlet steps corner to corner and skips your plaza.",
            "Everything lethal lives in the two wedges either side of it.",
            "The portals are the only fast way across. Take them or be caught.",
        ),
        outro=(
            "Light falls in behind you and does not come back out.",
            "One layer left.",
            "It has been waiting the whole time.",
        ),
        speaker="horizon",
    ),

    StoryBeat(
        level_index=11,
        chapter=4,
        chapter_title=_CH4,
        title="Everything, Refracted",
        intro=(
            "The core. Every hazard the machine still owns, all at once,",
            "arranged with terrible politeness.",
            "A cage of four beams, and only one face burns at a time.",
            "Enter on the dark face. Leave by the next.",
        ),
        outro=(
            "The cage opens. The core is small, and cold, and yours.",
            "The machine stops dying.",
            "It simply stops.",
        ),
        speaker="core",
    ),
)

BEAT_COUNT: int = len(BEATS)


# ==========================================================================
# Chapters
# ==========================================================================
CHAPTERS: Tuple[Chapter, ...] = (
    Chapter(
        number=1,
        title=_CH1,
        blurb=(
            "The outer lattice, half powered and mostly empty.",
            "Nothing here wants you dead yet.",
        ),
        first_index=0,
        last_index=2,
    ),
    Chapter(
        number=2,
        title=_CH2,
        blurb=(
            "Down where the machine still runs hot.",
            "Everything moves on a schedule, and none of it is yours.",
        ),
        first_index=3,
        last_index=5,
    ),
    Chapter(
        number=3,
        title=_CH3,
        blurb=(
            "Storage, growth and combustion.",
            "The parts of itself the machine locked away, and why.",
        ),
        first_index=6,
        last_index=8,
    ),
    Chapter(
        number=4,
        title=_CH4,
        blurb=(
            "Out to the quiet edge, then inward, past the bend in the light,",
            "to the thing that dreamed all of this.",
        ),
        first_index=9,
        last_index=11,
    ),
)

CHAPTER_COUNT: int = len(CHAPTERS)


# ==========================================================================
# Epilogue
# ==========================================================================
EPILOGUE: StoryCard = StoryCard(
    title="After",
    lines=(
        "You are the machine now. Every corridor, every gate, every dead lane.",
        "There is nothing left in here to eat.",
        "Somewhere far off, a colder lattice comes up dark",
        "and waits for its own first light.",
    ),
    speaker="",
)


# ==========================================================================
# Accessors - all clamped, none of them raise
# ==========================================================================
def _clamp_index(level_index: int, hi: int) -> int:
    """
    Coerce anything at all to a valid 0..hi index.  Never raises.

    The infinities need their own arm: ``int(float("inf"))`` raises
    ``OverflowError``, which is neither a ``TypeError`` nor a ``ValueError``,
    so a bare two-exception guard here would let it straight through and out of
    every accessor built on top of this one.  ``+inf`` saturates at the last
    beat and ``-inf`` (like NaN, and like anything uncoercible) at the first.
    """
    try:
        if level_index != level_index:            # NaN, without importing math
            return 0
    except Exception:                             # pragma: no cover - exotic
        return 0
    try:
        idx = int(level_index)
    except OverflowError:
        # Only the infinities land here, and only one of them is positive.
        try:
            return hi if level_index > 0 else 0
        except Exception:                         # pragma: no cover - exotic
            return 0
    except (TypeError, ValueError, AttributeError):
        return 0
    if idx < 0:
        return 0
    if idx > hi:
        return hi
    return idx


def get_beat(level_index: int) -> StoryBeat:
    """
    The beat for level `level_index`, clamped into range.

    Never raises and never returns ``None``: out-of-range or nonsense input
    yields the first or last beat, so a scene can call this unguarded.
    """
    return BEATS[_clamp_index(level_index, BEAT_COUNT - 1)]


def get_chapter(level_index: int) -> Chapter:
    """
    The chapter that owns level `level_index`, clamped into range.

    Never raises.  Derived from the beat rather than from arithmetic so the
    chapter grouping stays defined in exactly one place: the beat table.
    """
    number = get_beat(level_index).chapter
    for ch in CHAPTERS:
        if ch.number == number:
            return ch
    return CHAPTERS[_clamp_index(number - 1, CHAPTER_COUNT - 1)]


def beats_in_chapter(chapter_number: int) -> Tuple[StoryBeat, ...]:
    """Every beat in chapter `chapter_number` (1-based), in order. Never raises."""
    try:
        num = int(chapter_number)
    except (TypeError, ValueError, OverflowError, AttributeError):
        # OverflowError is the infinities; see _clamp_index for why it matters.
        return ()
    return tuple(b for b in BEATS if b.chapter == num)


def chapter_start(level_index: int) -> bool:
    """
    True when `level_index` is the first level of a chapter.

    This is the cue for the scene to show a chapter card before the level's own
    intro.  Clamped, so nonsense input answers for the nearest real level.
    """
    return get_chapter(level_index).first_index == \
        _clamp_index(level_index, BEAT_COUNT - 1)


def chapter_end(level_index: int) -> bool:
    """True when clearing `level_index` closes out its chapter."""
    return get_chapter(level_index).last_index == \
        _clamp_index(level_index, BEAT_COUNT - 1)


def all_beats() -> Tuple[StoryBeat, ...]:
    """The whole campaign narrative, in order."""
    return BEATS


# ==========================================================================
# Self-check
# --------------------------------------------------------------------------
# core.level is imported lazily, inside the function, so that importing this
# module stays free of every other dependency in the package.
# ==========================================================================
def validate_story() -> List[str]:
    """
    Cross-check the narrative against the level table and itself.

    Returns a list of human-readable problems; empty means consistent.  Run by
    ``python -m snake.core.story``.
    """
    problems: List[str] = []

    if BEAT_COUNT != CHAPTER_COUNT * CHAPTER_SIZE:
        problems.append("{} beats does not fill {} chapters of {}".format(
            BEAT_COUNT, CHAPTER_COUNT, CHAPTER_SIZE))

    for i, beat in enumerate(BEATS):
        if beat.level_index != i:
            problems.append("beat {}: level_index is {}".format(i, beat.level_index))
        if not beat.intro or not beat.outro:
            problems.append("{}: empty intro or outro".format(beat.title))
        if len(beat.intro) > 4 or len(beat.outro) > 4:
            problems.append("{}: more than four lines in a block".format(beat.title))
        ch = get_chapter(i)
        if beat.chapter_title != ch.title:
            problems.append("{}: chapter_title '{}' != '{}'".format(
                beat.title, beat.chapter_title, ch.title))
        if not ch.contains(i):
            problems.append("{}: outside chapter {} range".format(beat.title, ch.number))

    covered: List[int] = []
    for ch in CHAPTERS:
        if ch.last_index < ch.first_index:
            problems.append("chapter {}: inverted level range".format(ch.number))
        if not ch.blurb:
            problems.append("chapter {}: empty blurb".format(ch.number))
        covered.extend(ch.level_indices)
    if sorted(covered) != list(range(BEAT_COUNT)):
        problems.append("chapters do not cover every level exactly once")

    starts = [i for i in range(BEAT_COUNT) if chapter_start(i)]
    if starts != [ch.first_index for ch in CHAPTERS]:
        problems.append("chapter_start() disagrees with the chapter table: "
                        "{}".format(starts))

    if not PROLOGUE.lines or not EPILOGUE.lines:
        problems.append("prologue or epilogue has no lines")

    # -- agreement with the actual campaign --------------------------------
    try:
        from . import level as level_mod
    except Exception as exc:             # pragma: no cover - import guard only
        problems.append("could not import core.level: {}".format(exc))
        return problems

    if level_mod.LEVEL_COUNT != BEAT_COUNT:
        problems.append("{} levels but {} beats".format(
            level_mod.LEVEL_COUNT, BEAT_COUNT))
    expected = ("Neon Grid", "Deep Nebula", "Emerald Circuit", "Solar Flare",
                "Abyssal Tide", "Violet Static", "Frozen Vault", "Toxic Bloom",
                "Crimson Engine", "Aurora Drift", "Event Horizon", "Prism Core")
    for i, name in enumerate(expected[:level_mod.LEVEL_COUNT]):
        if level_mod.get_level(i).name != name:
            problems.append("level {} is '{}', story was written for '{}'".format(
                i + 1, level_mod.get_level(i).name, name))
    return problems


if __name__ == "__main__":               # pragma: no cover - developer utility
    print("PROLOGUE - {}".format(PROLOGUE.title))
    for _line in PROLOGUE.lines:
        print("    " + _line)
    for _ch in CHAPTERS:
        print("\n{}  (levels {}-{})".format(_ch, _ch.first_index + 1,
                                            _ch.last_index + 1))
        for _line in _ch.blurb:
            print("    " + _line)
        for _b in beats_in_chapter(_ch.number):
            print("  {:02d} {:<32} [{}]".format(_b.number, _b.title, _b.speaker))
            for _line in _b.intro:
                print("       > " + _line)
            for _line in _b.outro:
                print("       < " + _line)
    print("\nEPILOGUE - {}".format(EPILOGUE.title))
    for _line in EPILOGUE.lines:
        print("    " + _line)
    _issues = validate_story()
    print("\n{} problem(s)".format(len(_issues)))
    for _p in _issues:
        print("  ! " + _p)

File: web/src/core/story.ts
/**
 * Story mode narrative for NEON SERPENT - a port of `snake/core/story.py`.
 *
 * The campaign is a descent. A lone data-serpent wakes on the outer lattice of a
 * dying machine and eats its way inward, layer by layer, until it reaches the
 * core that dreamed the whole thing. Twelve levels, four chapters of three, one
 * beat per level.
 *
 * This module is **pure data**: no rendering, no state, no I/O. It knows nothing
 * about scenes; it only hands out text. Every accessor clamps its input and is
 * total - it never throws and never returns `undefined` - so a scene can call
 * `getBeat(game.levelIndex)` unguarded. NaN, the infinities and non-integers
 * each have an explicit arm: the Python original had a real bug where
 * `int(float("inf"))` raised `OverflowError` straight out of every accessor
 * built on `_clamp_index`, so the clamping here saturates instead.
 *
 * The text itself lives in `data/story.json`, exported from the Python module.
 */

import rawStory from "../data/story.json";
import rawLevels from "../data/levels.json";

// ==========================================================================
// Shapes
// ==========================================================================

/** Levels per chapter. Twelve levels / four chapters. */
export const CHAPTER_SIZE: number = 3;

/**
 * One level's worth of narrative.
 */
export interface StoryBeat {
  /** Zero-based level this beat belongs to (matches `LEVELS[levelIndex]`). */
  readonly levelIndex: number;
  /** One-based chapter number, 1..4. */
  readonly chapter: number;
  /**
   * Title of the chapter this beat sits in, repeated here so a caller with only
   * a beat never has to look the chapter up.
   */
  readonly chapterTitle: string;
  /** The beat's own title, shown on the card above the lines. */
  readonly title: string;
  /** Short lines shown *before* the level is played. */
  readonly intro: readonly string[];
  /** Short lines shown *after* the level is cleared. */
  readonly outro: readonly string[];
  /** Flavour label for who or what is talking; `""` means unattributed. */
  readonly speaker: string;
}

/**
 * A standalone narrative card that belongs to no single level.
 *
 * Used for the prologue (once, when a new story run starts) and the epilogue
 * (after Prism Core falls). Same shape as a beat's text block so a scene can
 * render all three with one routine.
 */
export interface StoryCard {
  /** Card title, shown large. */
  readonly title: string;
  /** The card's lines, in order. */
  readonly lines: readonly string[];
  /** Flavour label for the speaker; `""` means unattributed. */
  readonly speaker: string;
}

/**
 * One act of the descent: three consecutive levels sharing a mood.
 */
export interface Chapter {
  /** One-based chapter number, 1..4. */
  readonly number: number;
  /** Chapter title, shown large on the chapter card. */
  readonly title: string;
  /** One or two lines setting the mood for the whole act. */
  readonly blurb: readonly string[];
  /** Zero-based index of the chapter's first level. */
  readonly firstIndex: number;
  /** Zero-based index of the chapter's last level (inclusive). */
  readonly lastIndex: number;
  /** The chapter number as a roman numeral, for the card header. */
  roman(): string;
  /** `[firstIndex, lastIndex]`, both zero-based and inclusive. */
  levelRange(): [number, number];
  /** Every zero-based level index in this chapter, in order. */
  levelIndices(): number[];
  /** True when `levelIndex` falls inside this chapter. Never throws. */
  contains(levelIndex: number): boolean;
}

const ROMAN: readonly string[] = ["", "I", "II", "III", "IV", "V", "VI"];

/** Concrete {@link Chapter}, so the derived views stay methods and not stored data. */
class ChapterRecord implements Chapter {
  constructor(
    readonly number: number,
    readonly title: string,
    readonly blurb: readonly string[],
    readonly firstIndex: number,
    readonly lastIndex: number,
  ) {}

  roman(): string {
    if (Number.isInteger(this.number) && this.number > 0 && this.number < ROMAN.length) {
      return ROMAN[this.number] ?? String(this.number);
    }
    return String(this.number);
  }

  levelRange(): [number, number] {
    return [this.firstIndex, this.lastIndex];
  }

  levelIndices(): number[] {
    const out: number[] = [];
    if (!Number.isFinite(this.firstIndex) || !Number.isFinite(this.lastIndex)) return out;
    for (let i = Math.trunc(this.firstIndex); i <= Math.trunc(this.lastIndex); i++) {
      out.push(i);
    }
    return out;
  }

  contains(levelIndex: number): boolean {
    const idx = typeof levelIndex === "number" ? levelIndex : Number(levelIndex);
    if (!Number.isFinite(idx)) return false;
    const i = Math.trunc(idx);
    return this.firstIndex <= i && i <= this.lastIndex;
  }
}

// ==========================================================================
// Loading the exported narrative
// ==========================================================================

interface RawBeat {
  readonly level_index?: unknown;
  readonly chapter?: unknown;
  readonly chapter_title?: unknown;
  readonly title?: unknown;
  readonly intro?: unknown;
  readonly outro?: unknown;
  readonly speaker?: unknown;
}

interface RawChapter {
  readonly number?: unknown;
  readonly title?: unknown;
  readonly blurb?: unknown;
  readonly first_index?: unknown;
  readonly last_index?: unknown;
}

interface RawCard {
  readonly title?: unknown;
  readonly lines?: unknown;
  readonly speaker?: unknown;
}

const RAW = rawStory as unknown as {
  beats?: readonly RawBeat[];
  chapters?: readonly RawChapter[];
  prologue?: RawCard;
  epilogue?: RawCard;
};

function intOf(v: unknown, fallback: number): number {
  const n = typeof v === "number" ? v : Number(v);
  return Number.isFinite(n) ? Math.trunc(n) : fallback;
}

function textOf(v: unknown, fallback = ""): string {
  return typeof v === "string" ? v : fallback;
}

function linesOf(v: unknown): readonly string[] {
  if (!Array.isArray(v)) return [];
  return (v as unknown[]).filter((s): s is string => typeof s === "string");
}

function cardOf(v: RawCard | undefined): StoryCard {
  return {
    title: textOf(v?.title),
    lines: linesOf(v?.lines),
    speaker: textOf(v?.speaker),
  };
}

/** The twelve beats, in level order. */
export const BEATS: readonly StoryBeat[] = (RAW.beats ?? []).map((b, i) => ({
  levelIndex: intOf(b.level_index, i),
  chapter: intOf(b.chapter, Math.floor(i / CHAPTER_SIZE) + 1),
  chapterTitle: textOf(b.chapter_title),
  title: textOf(b.title),
  intro: linesOf(b.intro),
  outro: linesOf(b.outro),
  speaker: textOf(b.speaker),
}));

/** How many beats the campaign has. */
export const BEAT_COUNT: number = BEATS.length;

/** The four chapters, in order. */
export const CHAPTERS: readonly Chapter[] = (RAW.chapters ?? []).map(
  (c, i) =>
    new ChapterRecord(
      intOf(c.number, i + 1),
      textOf(c.title),
      linesOf(c.blurb),
      intOf(c.first_index, i * CHAPTER_SIZE),
      intOf(c.last_index, i * CHAPTER_SIZE + CHAPTER_SIZE - 1),
    ),
);

/** How many chapters the campaign has. */
export const CHAPTER_COUNT: number = CHAPTERS.length;

/** Shown once, when a new story run starts. */
export const PROLOGUE: StoryCard = cardOf(RAW.prologue);

/** Shown after Prism Core falls. */
export const EPILOGUE: StoryCard = cardOf(RAW.epilogue);

/** The beat handed back when the table is empty; keeps every accessor total. */
const EMPTY_BEAT: StoryBeat = {
  levelIndex: 0,
  chapter: 1,
  chapterTitle: "",
  title: "",
  intro: [],
  outro: [],
  speaker: "",
};

/** The chapter handed back when the table is empty; keeps every accessor total. */
const EMPTY_CHAPTER: Chapter = new ChapterRecord(1, "", [], 0, 0);

// ==========================================================================
// Accessors - all clamped, none of them throw
// ==========================================================================

/**
 * Coerce anything at all to a valid `0..hi` index. Never throws.
 *
 * `+Infinity` saturates at `hi`; `-Infinity`, `NaN` and anything uncoercible
 * land at `0`. Non-integers truncate toward zero, matching Python's `int()`.
 */
function clampIndex(levelIndex: number, hi: number): number {
  const raw = typeof levelIndex === "number" ? levelIndex : Number(levelIndex);
  if (Number.isNaN(raw)) return 0;
  const top = hi > 0 ? hi : 0;
  if (raw === Number.POSITIVE_INFINITY) return top;
  if (raw === Number.NEGATIVE_INFINITY) return 0;
  if (!Number.isFinite(raw)) return 0;
  const idx = Math.trunc(raw);
  if (idx < 0) return 0;
  if (idx > top) return top;
  return idx;
}

/**
 * The beat for level `levelIndex`, clamped into range.
 *
 * Never throws and never returns `undefined`: out-of-range or nonsense input
 * yields the first or last beat, so a scene can call this unguarded.
 */
export function getBeat(levelIndex: number): StoryBeat {
  if (BEAT_COUNT === 0) return EMPTY_BEAT;
  return BEATS[clampIndex(levelIndex, BEAT_COUNT - 1)] ?? EMPTY_BEAT;
}

/**
 * The chapter that owns level `levelIndex`, clamped into range.
 *
 * Never throws. Derived from the beat rather than from arithmetic so the
 * chapter grouping stays defined in exactly one place: the beat table.
 */
export function getChapter(levelIndex: number): Chapter {
  if (CHAPTER_COUNT === 0) return EMPTY_CHAPTER;
  const number = getBeat(levelIndex).chapter;
  for (const ch of CHAPTERS) {
    if (ch.number === number) return ch;
  }
  return CHAPTERS[clampIndex(number - 1, CHAPTER_COUNT - 1)] ?? EMPTY_CHAPTER;
}

/**
 * Every beat in chapter `chapterNumber` (1-based), in order.
 *
 * Never throws; NaN, the infinities and anything uncoercible give an empty
 * array, and a non-integer truncates toward zero before matching.
 */
export function beatsInChapter(chapterNumber: number): StoryBeat[] {
  const raw = typeof chapterNumber === "number" ? chapterNumber : Number(chapterNumber);
  if (!Number.isFinite(raw)) return [];
  const num = Math.trunc(raw);
  return BEATS.filter((b) => b.chapter === num);
}

/**
 * True when `levelIndex` is the first level of a chapter.
 *
 * This is the cue for the scene to show a chapter card before the level's own
 * intro. Clamped, so nonsense input answers for the nearest real level.
 */
export function chapterStart(levelIndex: number): boolean {
  if (BEAT_COUNT === 0) return false;
  return getChapter(levelIndex).firstIndex === clampIndex(levelIndex, BEAT_COUNT - 1);
}

/** True when clearing `levelIndex` closes out its chapter. */
export function chapterEnd(levelIndex: number): boolean {
  if (BEAT_COUNT === 0) return false;
  return getChapter(levelIndex).lastIndex === clampIndex(levelIndex, BEAT_COUNT - 1);
}

/** The whole campaign narrative, in order. */
export function allBeats(): readonly StoryBeat[] {
  return BEATS;
}

// ==========================================================================
// Self-check
// ==========================================================================

/** The level names the narrative was written against, in order. */
const EXPECTED_LEVEL_NAMES: readonly string[] = [
  "Neon Grid", "Deep Nebula", "Emerald Circuit", "Solar Flare",
  "Abyssal Tide", "Violet Static", "Frozen Vault", "Toxic Bloom",
  "Crimson Engine", "Aurora Drift", "Event Horizon", "Prism Core",
];

/**
 * Cross-check the narrative against the level table and itself.
 *
 * Returns a list of human-readable problems; empty means consistent. The level
 * cross-check reads `data/levels.json` directly rather than `core/level.ts`, so
 * this module stays free of every other dependency in the package.
 */
export function validateStory(): string[] {
  const problems: string[] = [];

  if (BEAT_COUNT !== CHAPTER_COUNT * CHAPTER_SIZE) {
    problems.push(
      `${BEAT_COUNT} beats does not fill ${CHAPTER_COUNT} chapters of ${CHAPTER_SIZE}`,
    );
  }

  for (let i = 0; i < BEAT_COUNT; i++) {
    const beat = BEATS[i];
    if (!beat) continue;
    if (beat.levelIndex !== i) problems.push(`beat ${i}: levelIndex is ${beat.levelIndex}`);
    if (beat.intro.length === 0 || beat.outro.length === 0) {
      problems.push(`${beat.title}: empty intro or outro`);
    }
    if (beat.intro.length > 4 || beat.outro.length > 4) {
      problems.push(`${beat.title}: more than four lines in a block`);
    }
    const ch = getChapter(i);
    if (beat.chapterTitle !== ch.title) {
      problems.push(`${beat.title}: chapterTitle '${beat.chapterTitle}' != '${ch.title}'`);
    }
    if (!ch.contains(i)) {
      problems.push(`${beat.title}: outside chapter ${ch.number} range`);
    }
  }

  const covered: number[] = [];
  for (const ch of CHAPTERS) {
    if (ch.lastIndex < ch.firstIndex) {
      problems.push(`chapter ${ch.number}: inverted level range`);
    }
    if (ch.blurb.length === 0) problems.push(`chapter ${ch.number}: empty blurb`);
    covered.push(...ch.levelIndices());
  }
  covered.sort((a, b) => a - b);
  const expectedCover = Array.from({ length: BEAT_COUNT }, (_, i) => i);
  if (covered.length !== expectedCover.length
      || covered.some((v, i) => v !== expectedCover[i])) {
    problems.push("chapters do not cover every level exactly once");
  }

  const starts: number[] = [];
  for (let i = 0; i < BEAT_COUNT; i++) if (chapterStart(i)) starts.push(i);
  const firsts = CHAPTERS.map((ch) => ch.firstIndex);
  if (starts.length !== firsts.length || starts.some((v, i) => v !== firsts[i])) {
    problems.push(`chapterStart() disagrees with the chapter table: ${starts.join(",")}`);
  }

  if (PROLOGUE.lines.length === 0 || EPILOGUE.lines.length === 0) {
    problems.push("prologue or epilogue has no lines");
  }

  // -- agreement with the actual campaign ----------------------------------
  const levels = (rawLevels as unknown as { levels?: readonly { name?: unknown }[] }).levels;
  if (!Array.isArray(levels)) {
    problems.push("could not read data/levels.json");
    return problems;
  }
  if (levels.length !== BEAT_COUNT) {
    problems.push(`${levels.length} levels but ${BEAT_COUNT} beats`);
  }
  const n = Math.min(levels.length, EXPECTED_LEVEL_NAMES.length);
  for (let i = 0; i < n; i++) {
    const got = textOf(levels[i]?.name);
    const want = EXPECTED_LEVEL_NAMES[i];
    if (got !== want) {
      problems.push(`level ${i + 1} is '${got}', story was written for '${want}'`);
    }
  }
  return problems;
}
